### DISCUSSION TITLE
SystemEntity and EntityManager

### DISCUSSION THREAD

**Discussion by MathiasPaulin** [STORM-IRIT/Radium-Engine#768] [Design discussion]
As mentioned in the PR #763 and the issue #767, The singleton `SystemEntity` is managed by the `EntityManager` and the lifetime of this singleton is the one of the `EntityManager`. 
The `EntityManager`, that is not a singleton, has the lifetime of the `Engine` and is fully managed by the `Engine`

This lifetime restriction make the notion of singleton quite unappropriated for `SystemEntity`.
Note that this discussion should also be related to other singleton whose lifetime is the one of the `Engine` : 
 - `ComponentMessenger`
 - `DebugRender`
 - `SystemEntity`

As all of these singletons are specific to the `Engine` instance (that is a singleton), we can change their status to private attribute of the engine, with methods in the `Engine` allowing to access these attributes. We already did this for `SignalManager`, `RenderObjectManager`, `ShaderManager` and `TextureManager`.

ComponentMessenger, DebugRender and SystemEntity, whose lifetime is the one of the Engine should be integrated into the Engine as the other managers.

If we do this, there will remain only 2 singletons in the Radium libs : 
 - the Engine (in the Radium::Engine library component)
 - the KeyMappingManager (in the Radium::Gui library component)
 
So the question is : 
 - if some objects have the lifetime of another object and should be used only within the constext defined by this other object, could they be singletons ?

I think they don't and propose to integrate the 3 singletons above as private attributes into the Engine.

What about this proposal ?

**Comment by nmellado** [comment]
I am fine with removing the singletons and make them as private attribute of `Engine`. 
We just need to be sure that if someone runs two Engines in the same app (I don't know why, but anyway), these attributes also work properly (ie. no static data structures).

**Reply by MathiasPaulin** [reply]
Engine is a singleton, so, there is no way to run two Engines in the same app, right ?
